### PR TITLE
Share SA private key with `master` Nodes

### LINKED ISSUE #1135: Control-plane expansion fails to deploy k-c-m
**Component**: salt

**What happened**:

When running the `deploy_node` orchestrate for a master node, it fails when installing k-c-m, since `/etc/kubernetes/pki/sa.key` is missing.

**What was expected**:

The Service Account private key is used by k-c-m to sign SA tokens, so it must me shared between master nodes.

The `deploy_node` orchestrate should not fail, and take care of having this key available where needed.

**Steps to reproduce**:

- `git checkout development/2.0`
- `./doit.sh vagrant_up` (let's go with a fresh one if possible)
- Start a new node:
    - `vagrant up node1`
- Retrieve its control-plane IP:
    - `vagrant ssh node1 -- sudo ip addr show eth0`
- Prepare a Node manifest for the new node:
    - `sed 's/<IP_ADDRESS>/"$node_ip"/' examples/new-node_vagrant.yaml > node_manifest.yaml`
- SSH into the bootstrap node:
    - `vagrant ssh bootstrap`
    - `sudo su`
    - `export KUBECONFIG=/etc/kubernetes/admin.conf`
- Create the Node object:
    - `kubectl apply -f /vagrant/node_manifest.yaml`
- Start the `deploy_new_node` orchestrate:
    - `kubectl exec salt-master-bootstrap -c salt-master -n kube-system -- salt-run state.orchestrate metalk8s.orchestrate.deploy_node saltenv=metalk8s-2.0 pillar='{"orchestrate": {"node_name": "node1"}}'`

The orchestrate should fail, with:

```
[...]
      ID: Create kube-controller-manager Pod manifest
Function: metalk8s.static_pod_managed
    Name: /etc/kubernetes/manifests/kube-controller-manager.yaml
  Result: False
 Comment: An exception occurred in this state: Traceback (most recent call last):
            File "/usr/lib/python2.7/site-packages/salt/state.py", line 1919, in call
              **cdata['kwargs'])
            File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1918, in wrapper
              return f(*args, **kwargs)
            File "/var/cache/salt/minion/extmods/states/metalk8s.py", line 19, in static_pod_managed
              for config_file in config_files or []
            File "/usr/lib/python2.7/site-packages/salt/modules/hashutil.py", line 75, in digest_file
              "File path '{0}' not found.".format(infile))
          CommandExecutionError: File path '/etc/kubernetes/pki/sa.key' not found.
 Started: 08:34:48.263226
Duration: 3.474 ms
 Changes:
[...]
```

**Resolution proposal** (optional):

Using a pillar for sharing this private data to only a subset of minions seems like the obvious approach.
Unless the master must not have access to `/etc/kubernetes/pki` (which it does for now), in which case we'd need to setup symmetric encryption between master nodes.

### PR DESCRIPTION
**Pull request by gdemonet**
**Component**: salt

**Context**: Control-plane expansion is broken since #1080, because we require the SA private key to be available where we deploy `k-c-m` static Pods.

**Summary**:

- Introduce an `ext_pillar` for passing secure data to a subset of minions (filtering is done using the Node roles)
- Use this information in `metalk8s.kubernetes.sa.advertised` to write the `sa.key` file if on a `master` Node
- Include `...sa.advertised` in `metalk8s.kubernetes.controller-manager.installed`

**Acceptance criteria**: 

- Control-plane expansion (follow the steps outlined in #1135) should work

---

Fixes: #1135

### CODE REVIEW

File: salt/metalk8s/kubernetes/controller-manager/installed.sls [hunk 1]
```diff
@@ -23,8 +25,6 @@ Create kube-controller-manager Pod manifest:
           - --address=127.0.0.1
           - --allocate-node-cidrs=true
           - --cluster-cidr={{ networks.pod }}
-          - --cluster-signing-cert-file=/etc/kubernetes/pki/ca.crt
-          - --cluster-signing-key-file=/etc/kubernetes/pki/ca.key
```
Comment (TeddyAndrieux) [inline salt/metalk8s/kubernetes/controller-manager/installed.sls:27]: Why remove these ?

File: buildchain/buildchain/salt_tree.py [hunk 2]
```diff
@@ -259,6 +259,7 @@ def task__deploy_salt_tree() -> Iterator[types.TaskDict]:
     Path('salt/_modules/metalk8s.py'),
 
     Path('salt/_pillar/metalk8s.py'),
+    Path('salt/_pillar/metalk8s_certs.py'),
```
Comment (TeddyAndrieux) [inline buildchain/buildchain/salt_tree.py]: Maybe rename the file and the variable and pillar key I think

**Review by TeddyAndrieux — DISMISSED** [review]

File: salt/_pillar/metalk8s_certs.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def _read_sa_private_key():
+    # FIXME: we only have access to this file because:
+    #   1- Salt master Pod mounts all of /etc/kubernetes
+    #   2- Salt master Pod runs on the 'bootstrap' minion, which is, by chance,
+    #      also the 'ca' minion, the owner of this key
+    # Suggestion: read through https://github.com/saltstack/salt/issues/45882
+    if not os.path.isfile(SA_PRIVATE_KEY_PATH):
+        log.warn("Missing SA key to share with master node.")
```
Comment (TeddyAndrieux) [inline salt/_pillar/metalk8s_certs.py]: Following #1018 you should maybe start not log directly but put those in `_errors` key but maybe not part of this PR as you want

Comment (gdemonet) [reply salt/_pillar/metalk8s_certs.py]: I'll also change the other log (skipping unknown minions) to `debug`

File: buildchain/buildchain/salt_tree.py (hunk 2 above)
Comment (gdemonet) [reply buildchain/buildchain/salt_tree.py]: I'll go for `metalk8s_private` if that's fine with you

File: salt/metalk8s/kubernetes/controller-manager/installed.sls (hunk 1 above)
Comment (gdemonet) [reply salt/metalk8s/kubernetes/controller-manager/installed.sls:27]: They were leftovers from the initial import of `kubeadm` generated manifests, which includes the [TLS Bootstrapping](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet-tls-bootstrapping/) set of features, that we don't want to use in MetalK8s.

The commit message explains briefly:

```
Fix k-c-m manifest

We remove the TLS bootstrap feature (--cluster-signing-... options),
and make sure k-c-m deployment is preceded by the creation of its
dependencies: the root CA cert and SA private key.
```

File: buildchain/buildchain/salt_tree.py (hunk 2 above)
Comment (TeddyAndrieux) [reply buildchain/buildchain/salt_tree.py]: Yes it's good

**Review by TeddyAndrieux — APPROVED** [review]
LGTM

**Comment by gdemonet** [comment]
@bert-e approve

**Comment by gdemonet** [comment]
@bert-e reset
